src/module/actor/character/crafting/entry.ts

```diff
@@ -22,18 +22,15 @@ export class CraftingEntry implements CraftingEntryData {
         this.actorPreparedFormulas = data.actorPreparedFormulas;
         this.selector = data.selector;
         this.name = data.name;
-        this.isAlchemical = data.isAlchemical ?? false;
-        this.isDailyPrep = data.isDailyPrep ?? false;
-        this.isPrepared = data.isPrepared ?? false;
+        this.isAlchemical = !!data.isAlchemical;
+        this.isDailyPrep = !!data.isDailyPrep;
+        this.isPrepared = !!data.isPrepared;
         this.maxSlots = data.maxSlots ?? 0;
         this.maxItemLevel = data.maxItemLevel || parentActor.level;
         this.fieldDiscovery = data.fieldDiscovery;
         this.batchSize = data.batchSize;
         this.fieldDiscoveryBatchSize = data.fieldDiscoveryBatchSize;
-
-        this.requiredTraits = data.requiredTraits ?? [[]];
-        if (this.requiredTraits.length === 0) this.requiredTraits.push([]);
-
+        this.craftableItems = new PredicatePF2e(data.craftableItems ?? {});
```

Comment:
If what you're saying is correct about persistence, you might not want to do new PredicatePF2e() here?

Reply:
Move it back to ```checkEntryRequirements``` ?

Reply:
Could maybe validate it using the method shark mentioned, which just got an upgrade to also be a type guard. I suggested to create a predicate here, but that was before I was aware it was persisted :(

Reply:
I'm already validating in the RE before passing to the entry

I have a seperate PR up to unpersist these entries so a lot of these issues will go away

Reply:
If so, then might best to draft this MR and do that one first.